drivers/neopixel/atmega.c

```diff
@@ -0,0 +1,206 @@
+     * the duration of the second phase (the low period) is in reality super
+     * lax. Therefore, we bit bang the high period precisely in inline assembly,
+     * and completely ignore the timing of the low phase. This works just fine.
+     *
+     * Further: For an 8MHz clock, we need a single CPU cycle access to the
+     * GPIO port. The is only possible with the `out` instruction, which takes
+     * the port to write to as immediate. Thus, it has to be known at compile
+     * time. We therefore simply provide 3 implementations for each of the
+     * three most commonly used GPIO ports.
+     *
+     * On final trick: A relative jump to the next instructions takes two
+     * CPU cycles; this, its 2 NOPs for the price of one (in terms of ROM
+     * consumption).
+     */
+    const uint8_t port_num = _port_num(dev->params.pin);
+    switch (port_num) {
+        case PORT_B:
+            while (pos < end) {
+                uint8_t cnt = 8;
+                uint8_t data = *pos;
+                while (cnt > 0) {
+                    __asm__ volatile(                       /* Cycles | 1 | 0 */
+                        "out    %[port], %[on]"     "\n\t"  /* 1      | x | x */
+                        "sbrs   %[data], 7"         "\n\t"  /* 1      | x | x */
+                        "out    %[port], %[off]"    "\n\t"  /* 1      | - | x */
+                        "rjmp   .+0"                "\n\t"  /* 2      | x | x */
+                        "out    %[port], %[off]"    "\n\t"  /* 1      | x | x */
+                        /* Total CPU cycles for high period:
+                         * 5 cycles for bit 1, 2 cycles for bit 0 */
+                        : [data]    "+r" (data)
+                        : [port]    "I" (_SFR_IO_ADDR(PORTB)),
+                          [on]      "r" (mask_on),
+                          [off]     "r" (mask_off)
+                    );
+                    cnt--;
+                    data <<= 1;
+                }
+                pos++;
+            }
+            break;
+        case PORT_C:
+            while (pos < end) {
+                uint8_t cnt = 8;
+                uint8_t data = *pos;
+                while (cnt > 0) {
+                    __asm__ volatile(                       /* Cycles | 1 | 0 */
+                        "out    %[port], %[on]"     "\n\t"  /* 1      | x | x */
+                        "sbrs   %[data], 7"         "\n\t"  /* 1      | x | x */
+                        "out    %[port], %[off]"    "\n\t"  /* 1      | - | x */
+                        "rjmp   .+0"                "\n\t"  /* 2      | x | x */
+                        "out    %[port], %[off]"    "\n\t"  /* 1      | x | x */
+                        /* Total CPU cycles for high period:
+                         * 5 cycles for bit 1, 2 cycles for bit 0 */
+                        : [data]    "+r" (data)
+                        : [port]    "I" (_SFR_IO_ADDR(PORTC)),
```

Comment:
Would it make sense to have a function that calculates `_SFR_IO_ADDR` based on `port_num` so you don't have to copy & paste this for every port?